Miscellaneous Travis improvements

vadz opened the pull request.

vadz committed:
Make Travis logs more readable by folding the different sections

See https://github.com/travis-ci/travis-ci/issues/2285 for more information
about log folding on Travis.

vadz committed:
Really check building using the installed version in Travis builds

The command to do this at the end of the script didn't actually do anything
because the minimal sample had been already built in tree before.

vadz committed:
Test builds with newer g++, clang and OS X on Travis CI

Configure the build matrix to test building on more platforms and using clang
as well and also test C++11 builds using both g++ and clang.

vadz committed:
Extract Travis before install script steps into a separate file

This allows to do make these steps more complicated, e.g. avoid using apt-get
under non-Linux (and even non-Debian/Ubuntu) platforms.

Also install GTK+ development package explicitly, while it's already available
on the older Travis Ubuntu 12.04 VMs, this is not the case for the newer 14.04
ones.

tierra commented:
Nice. I've been wanting to get OSX in the mix here for a while. Same with some trusty builds. Hopefully you have some good luck with this.

vadz commented:
It mostly seems to work fine, but I just can't convince homebrew to install the universal (or even just 64 bit) version of cppunit somehow... I must be missing something because, if anything, it should install the 64 bit version _only_, but I see no other explanation: there is `-L/usr/local/Cellar/cppunit/1.13.2/lib -lcppunit` in the linker command line yet all cppunit symbols are not found as can be seen [in the log](https://travis-ci.org/wxWidgets/wxWidgets/jobs/99520780)

tierra commented on build/tools/before_install.sh:
```diff
@@ -0,0 +1,21 @@
+
+    Darwin)
+        brew update
+        brew install --universal cppunit
```
I think this flag is required to come *after* the package, so:
```brew install cppunit --universal```

vadz commented:
Unfortunately putting `--universal` at the end didn't help neither.

vadz commented:
I suspect the problem with cppunit was actually due to using the wrong C++ library due to the wrong use of 10.5 as minimal OS X version and this was fixed in c3504663efad901677155e484f2b00fa0fce7fc8, so hopefully everything should work now. Let's see...

vadz committed:
Pass -std=c++11 when using Unix makefile on Travis explicitly

wx-config doesn't include -std=c++11 in its output currently and it's not
clear whether it should (consider c++11 vs gnu++11 or even vs c++14), so add
this compiler option explicitly in C++11 Travis builds.
